pkg/jobWorker/job.go

```diff
@@ -0,0 +1,307 @@
+// The user running Start() should be the root user or have the necessary permissions to create namespaces and control groups
+//
+// ErrJobAlreadyStarted is returned, if the Job has already been started.
+// ErrInvalidCommand, ErrInvalidCPU, ErrInvalidIOBytesPerSecond, ErrInvalidMemBytes is returned, if provided configuration is invalid
+func (job *Job) Start() error {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	// validate configuration
+	log.Printf("validate job:%s", job)
+	if err := job.config.isValid(); err != nil {
+		log.Printf("validate job error:%v", err)
+		return err
+	}
+
+	// validate job hasn't been run
+	if job.status.State != JOB_STATUS_NOT_STARTED {
+		return ErrJobAlreadyStarted
+	}
+
+	cmd := exec.Command(job.config.Command, job.config.Arguments...)
+	// combine the stdout and stderr so that the stdout and stderr are combined in the order they are written
+	cmd.Stderr = job.output
+	cmd.Stdout = job.output
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		// CLONE_NEWPID:  creates a new PID namespace preventing the process from seeing/killing host processes
+		// CLONE_NEWNET:  creates a new network namespace preventing the process from accessing the internet or local network
+		// CLONE_NEWNS:   creates a new mount namespace preventing the process from impacting host mounts
+		// CLONE_NEWUTS:  creates a new UTS namespaces provide isolation between two system identifiers: the hostname and the NIS domain name
+		// CLONE_NEWPID:  crates new PID namespaces isolate the process ID number space, meaning that processes in different PID namespaces can have the same PID
+		// CLONE_NEWUSER: creates new namespaces to isolate security-related identifiers and attributes, in particular, user IDs and group IDs
+		Cloneflags: syscall.CLONE_NEWNS |
+			syscall.CLONE_NEWIPC |
+			syscall.CLONE_NEWNET |
+			syscall.CLONE_NEWUTS |
+			syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWUSER,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+		// Also, enables mounting a new proc filesystem so that command such as `ps -ef` only see the processes in the PID namespace
+		Unshareflags: syscall.CLONE_NEWNS,
+		// instruct cmd.Run to use the control group file descriptor, so that Job Command does not
+		// have to manually add the new PID to the control group
+		UseCgroupFD: true,
+	}
+
+	formatedUUID := strings.Replace(job.UUID.String(), "-", "", -1)
+
+	cgroupName := formatedUUID
+	err := ns.CreateCGroup(cgroupName)
+	if err != nil {
+		return fmt.Errorf("error creating cgroup: %w", err)
+	}
+
+	if err = ns.AddResourceControl(cgroupName, ns.CPU_WEIGHT_File, strconv.Itoa(int(job.config.CPU*100))); err != nil {
+		log.Printf("could not add resources into controller:%s, %v", ns.CPU_WEIGHT_File, err)
```

Comment:
leftover cgroups on these failure cases.

Reply:
@greedy52 good point. I made appropriate changes.
<img width="811" alt="Screenshot 2024-09-04 at 11 03 45 AM" src="https://github.com/user-attachments/assets/8d63de6e-1df9-4a5f-bad6-b04bdc7ddada">


here is testing
![Screenshot 2024-09-04 at 11 01 27 AM](https://github.com/user-attachments/assets/d67f9709-49bb-4590-b33e-968f44f50b42)